sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/multivolume/SevenZipRarMultiVolumeCallbackHandler.java

```diff
@@ -0,0 +1,89 @@
+package com.mucommander.sevenzipjbindings.multivolume;
+
+import com.mucommander.commons.file.AbstractFile;
+import com.mucommander.commons.file.FileFactory;
+import com.mucommander.commons.file.FileURL;
+import com.mucommander.sevenzipjbindings.SignatureCheckedRandomAccessFile;
+import net.sf.sevenzipjbinding.*;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SevenZipRarMultiVolumeCallbackHandler implements
+        IArchiveOpenVolumeCallback, IArchiveOpenCallback, ICryptoGetTextPassword, Closeable {
+
+    private Map<String, SignatureCheckedRandomAccessFile> fileCache = new HashMap<>();
+
+    private String lastFileName;
+
+    private final byte[] signature;
+
+    private final String password;
+
+    public SevenZipRarMultiVolumeCallbackHandler(byte[] signature, String password) {
+        this.signature = signature;
+        this.password = password;
+    }
+
+    @Override
+    public void setTotal(Long files, Long bytes) throws SevenZipException {
+        // NO-OP
+    }
+
+    @Override
+    public void setCompleted(Long files, Long bytes) throws SevenZipException {
+        // NO-OP
+    }
+
+    @Override
+    public Object getProperty(PropID propID) throws SevenZipException {
+        switch (propID) {
+            case NAME:
+                return lastFileName;
+        }
+        return null;
+    }
+
+    @Override
+    public IInStream getStream(String filename) throws SevenZipException {
+        try {
+            SignatureCheckedRandomAccessFile stream = fileCache.get(filename);
+            if (stream != null) {
+                stream.seek(0, ISeekableStream.SEEK_SET);
+            } else {
+                File tempFile = new File(filename);
+                AbstractFile abstractFile = FileFactory.createRawFile(FileURL.getFileURL(tempFile.getAbsolutePath()), null, new HashMap<>());
```

Comment:
why not to use FileFactory#getFile?

Reply:
That's the thing - using this method takes me into an endless recursion. Look at the stack trace:
```2023-04-21 20:17:38.566 [FelixStartLevel] ERROR com.mucommander.Application:439 - Startup failed
java.lang.StackOverflowError: null
	at java.base/java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3966)
	at java.base/java.util.regex.Pattern$Curly.match(Pattern.java:4375)
	at java.base/java.util.regex.Pattern$Curly.match0(Pattern.java:4395)
	at java.base/java.util.regex.Pattern$Curly.match(Pattern.java:4382)
	at java.base/java.util.regex.Pattern$Begin.match(Pattern.java:3686)
	at java.base/java.util.regex.Matcher.match(Matcher.java:1756)
	at java.base/java.util.regex.Matcher.matches(Matcher.java:713)
	at com.mucommander.commons.file.protocol.local.LocalFile.isRoot(LocalFile.java:874)
	at com.mucommander.commons.file.protocol.local.LocalFile.getName(LocalFile.java:762)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:82)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
... keeps on going...```

Reply:
That's why I needed to "open up" `FileFactory::createRawFile` which was `private`

Reply:
I tried it and it worked! I restored `FileFactory::createRawFile` access modifier back to `private` as well